crates/scmi/src/scmi.rs

```diff
@@ -0,0 +1,1377 @@
+        },
+    );
+    handlers.bind(
+        BASE_PROTOCOL_ID,
+        BASE_MESSAGE_ATTRIBUTES,
+        "base/message_attributes",
+        vec![ParameterType::UnsignedInt32],
+        message_attributes,
+    );
+    handlers.bind(
+        BASE_PROTOCOL_ID,
+        BASE_DISCOVER_VENDOR,
+        "base/discover_vendor",
+        vec![],
+        |_, _| -> Response {
+            Response::from_value(MessageValue::String(
+                "rust-vmm".to_string(),
+                MAX_SIMPLE_STRING_LENGTH,
+            ))
+        },
+    );
+    handlers.bind(
+        BASE_PROTOCOL_ID,
+        BASE_DISCOVER_IMPLEMENTATION_VERSION,
+        "base/discover_implementation_version",
+        vec![],
+        |_, _| -> Response { Response::from_value(MessageValue::Unsigned(0)) },
+    );
+    handlers.bind(
+        BASE_PROTOCOL_ID,
+        BASE_DISCOVER_LIST_PROTOCOLS,
+        "base/discover_list_protocols",
+        vec![ParameterType::UnsignedInt32],
+        discover_list_protocols,
+    );
+
+    fn discover_list_protocols(handler: &ScmiHandler, request: &ScmiRequest) -> Response {
+        // Base protocol is skipped
+        let skip: usize = request
+            .get_unsigned(0)
+            .try_into()
+            .expect("Extremely many protocols");
+        let protocols: Vec<ProtocolId> = handler
+            .handlers
+            .keys()
+            .filter(|(protocol_id, _)| *protocol_id != BASE_PROTOCOL_ID)
+            .map(|(protocol_id, _)| *protocol_id)
+            .unique()
+            .sorted()
+            .skip(skip)
+            .collect();
+        let n_protocols = protocols.len();
+        debug!("Number of listed protocols after {}: {}", skip, n_protocols);
+        let mut values: Vec<MessageValue> = vec![MessageValue::Unsigned(n_protocols as u32)];
+        if n_protocols > 0 {
+            let mut compressed: Vec<u32> = vec![0; 1 + (n_protocols - 1) / 4];
+            for i in 0..n_protocols {
+                debug!("Adding protocol: {}", protocols[i]);
+                compressed[i % 4] |= ((protocols[i]) as u32) << ((i % 4) * 8);
+            }
+            for item in compressed {
+                values.push(MessageValue::Unsigned(item));
+            }
+        }
+        Response::from_values(&mut values)
+    }
+}
+
+pub(crate) const SENSOR_PROTOCOL_ID: ProtocolId = 0x15;
+const SENSOR_VERSION: MessageId = 0x0;
+const SENSOR_ATTRIBUTES: MessageId = 0x1;
+const SENSOR_MESSAGE_ATTRIBUTES: MessageId = 0x2;
+pub(crate) const SENSOR_DESCRIPTION_GET: MessageId = 0x3;
+pub(crate) const SENSOR_READING_GET: MessageId = 0x6;
+pub(crate) const SENSOR_AXIS_DESCRIPTION_GET: MessageId = 0x7;
+pub(crate) const SENSOR_CONFIG_GET: MessageId = 0x9;
+pub(crate) const SENSOR_CONFIG_SET: MessageId = 0xA;
+pub(crate) const SENSOR_CONTINUOUS_UPDATE_NOTIFY: MessageId = 0xB;
+
+#[allow(dead_code)]
+pub(crate) const SENSOR_UNIT_NONE: u8 = 0;
```

Comment:
Should these units be an enum?

Reply:
I didn't use enum to avoid annoying type conversions.